smelter-core/src/pipeline/webrtc/whep_input.rs

```diff
@@ -0,0 +1,188 @@
+use std::sync::Arc;
+use std::thread;
+use std::time::{Duration, Instant};
+
+use crate::pipeline::{
+    input::Input,
+    rtp::RtpNtpSyncPoint,
+    webrtc::{
+        http_client::WhipWhepHttpClient,
+        peer_connection_recvonly::RecvonlyPeerConnection,
+        whep_input::{
+            establish_peer_connection::exchange_sdp_offers,
+            process_tracks::{process_audio_track, process_video_track},
+            resolve_video_preferences::resolve_video_preferences,
+        },
+    },
+};
+use crate::queue::QueueDataReceiver;
+use crossbeam_channel::{Sender, bounded};
+use tokio::sync::oneshot;
+use tracing::{Instrument, Level, debug, span, warn};
+use url::Url;
+use webrtc::rtp_transceiver::rtp_codec::RTPCodecType;
+
+use crate::prelude::*;
+
+mod establish_peer_connection;
+mod process_tracks;
+mod resolve_video_preferences;
+
+const WHEP_INIT_TIMEOUT: Duration = Duration::from_secs(60);
+
+#[derive(Debug)]
+pub struct WhepInput {
+    ctx: Arc<PipelineCtx>,
+    session_url: Url,
+    client: Arc<WhipWhepHttpClient>,
+}
+
+impl WhepInput {
+    pub(crate) fn new_input(
+        ctx: Arc<PipelineCtx>,
+        input_id: InputId,
+        options: WhepInputOptions,
+    ) -> Result<(Input, InputInitInfo, QueueDataReceiver), InputInitError> {
+        let (init_confirmation_sender, init_confirmation_receiver) = oneshot::channel();
+
+        let (frame_sender, frame_receiver) = bounded(5);
+        let (input_samples_sender, input_samples_receiver) = bounded(5);
+
+        let span = span!(
+            Level::INFO,
+            "WHEP client task",
+            input_id = input_id.to_string()
+        );
+        let rt = ctx.tokio_rt.clone();
+        let ctx_clone = ctx.clone();
+        rt.spawn(
+            async {
+                let result =
+                    init_whep_client(ctx_clone, options, input_samples_sender, frame_sender).await;
+                match result {
+                    Ok(handle) => {
+                        init_confirmation_sender.send(Ok(handle)).unwrap();
+                    }
+                    Err(err) => init_confirmation_sender.send(Err(err)).unwrap(),
+                }
+            }
+            .instrument(span),
+        );
+
+        let (session_url, client) =
+            wait_with_deadline(init_confirmation_receiver, WHEP_INIT_TIMEOUT)?;
+        Ok((
```

Comment:
Either:
- This function should have absolute minimum of logic, `init_whep_client` should already return the same type as new_input.
- Logic involving returning result by channels and waiting for deadline should not be in top level code like that